mitodl/ocw-studio
# Upload a transcoded video to youtube

Depends on #486 , but you should be able to test this by creating your own Youtube API project as described in the new README section.

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes #480

#### What's this PR do?
When AWS MediaConvert transcoding is complete, upload the appropriate video file to Youtube.

#### How should this be manually tested?
- Create your own youtube api configuration following the steps in the new README section.
- Set up Google Drive integration and transcoding as described for PR #469.  You will need to simulate the webhook endpoint request so that the required`VideoFile` objects will be created and the youtube upload task triggered.  The request body should contain this (update to match your settings/video/drive file):
   ```
   {
	"version": "0",
	"id": "426c759f-277e-ee0f-aceb-7fb79214bc51",
	"detail-type": "MediaConvert Job State Change",
	"source": "aws.mediaconvert",
	"account": "<settings.AWS_ACCOUNT_ID>"
	",
	"detail": {
		"timestamp": 1629911639065,
		"accountId": "919801701561",
		"queue": "arn:aws:mediaconvert:us-east-1:919801701561:queues/Default",
		"jobId": "<VideoJob.job_id>",
		"status": "COMPLETE",
		"userMetadata": {},
		"outputGroupDetails": [{
			"outputDetails": [{
					"outputFilePaths": [
						"s3://<settings.AWS_STORAGE_BUCKET_NAME>/aws_mediaconvert_transcodes/<Website.short_id>/<DriveFile.file_id>/<original_video_filename_base>_youtube.mp4"
					],
					"durationInMs": 45466,
					"videoDetails": {
						"widthInPx": 320,
						"heightInPx": 176
					}
				},
				{
					"outputFilePaths": [
						"s3://<settings.AWS_STORAGE_BUCKET_NAME>/aws_mediaconvert_transcodes/<Website.short_id>/<DriveFile.file_id>/<original_video_filename_base>_360p_16_9.mp4"
					],
					"durationInMs": 45466,
					"videoDetails": {
						"widthInPx": 640,
						"heightInPx": 360
					}
				},
				{
					"outputFilePaths": [
						"s3://<settings.AWS_STORAGE_BUCKET_NAME>/aws_mediaconvert_transcodes/<Website.short_id>/<DriveFile.file_id>/<original_video_filename_base>_360p_4_3.mp4"
					],
					"durationInMs": 45466,
					"videoDetails": {
						"widthInPx": 480,
						"heightInPx": 360
					}
				}
			],
			"type": "FILE_GROUP"
		}]
	}
  }
  ```
- Wait a few minutes, the `upload_youtube_videos` and `update_youtube_statuses` tasks should run.  Go to ../admin/videos/video/<video_id>/change/  and check that the `VideoFile` with `destination` of `youtube` now has a `destination_id` and a desination_status of either `uploaded` (initially) or `processed` (ultimately).
Once the `destination_status` is `processed`, the `status` should be `Complete`.
- Go to `https://www.youtube.com/watch?v=<destination_id>`, your video should play from there.


#### Any background context you want to provide?
Much of this code was originally from OVS and modified as necessary.
